src/EnergyPlus/BaseboardRadiator.cc

```diff
@@ -1230,8 +1230,8 @@ namespace BaseboardRadiator {
     }
 
     Real64 HWBaseboardUAResidual(EnergyPlusData &state,
-                                 Real64 const UA,                 // UA of coil
-                                 std::array<Real64, 2> const &Par // par(1) = design coil load [W]
+                                 Real64 const UA,                // UA of coil
+                                 std::array<Real64, 2> const Par // par(1) = design coil load [W]
```

Comment:
I'm surprised by this one. Is this correct?

Reply:
I'm surprised by this one too. From a performance standpoint, I think the argument is the same as the one for passing `std::string_view` and `gsl::span` by value. Each of those is a structure with two elements and passing two elements by value is a bit cheaper than passing the struct that contains them by reference (it's two instructions in the caller vs. one in the caller and two in the callee, so you are saving one instruction). However, despite this I do think it's better to use a reference in this case for the purpose of keeping the overall pattern simple, i.e., `std::string_view` and `gsl::span` by value and every other object by const reference. There is also the thing that `std::string_view` and `gsl::span` are themselves both essentially references and so passing a reference by reference is logically redundant, whereas that redundancy does not exist for random arrays of two elements.

Reply:
You have to remember that it's the first time I write a clang-tidy script, or even get exposed with the clang AST. I think I should have used a threshold of <= 15 bytes instead of <= 16 bytes when I applied it as well...

This is something I have done on my spare time (to address a specific OpenStudio issue that I could have addressed manually). I thought I'd try it out on E+ after mentioning it to Edwin, and I kinda messed this one up when applying to it took me like a half day to clean up manually, and I probably have missed a few things. This is just me nerdsnipping myself real bad (this isn't part of my scope so you won't find a pivotal ticket associated).

As far as the differences in output, I think some of the double/int arguments to the Serializing functions should be passed by const ref rather than value to avoid floating point errors/rounding...


I'm unfortunately short for free time to fix it for the next few weeks as I'm on paternity leave. I can close this issue or convert it to draft and we can revisit later?

Reply:
Hey, I'm not knocking it - this is great! I've wanted to figure out how to use the clang AST tools for some time. Maybe you can give me a lesson when you're back 😉 .

Go be a dad and don't worry about this. I think one of us can clean up/review/merge this and get it wrapped up.

Congrats!

Reply:
Congrats Julien! 🥳 👶

Reply:
> You have to remember that it's the first time I write a clang-tidy script, or even get exposed with the clang AST. I think I should have used a threshold of <= 15 bytes instead of <= 16 bytes when I applied it as well...

The safe threshold is actually <= 8 bytes, but I'm not sure that looking only at argument size is enough. If you pass a struct/class with two 32-bit members or an array of 8 8-bit members by value, the elements would not be packed into a single register, they would be splayed out over multiple registers (I think). You wouldn't be able to access the members directly out of a single register. The safe check should be <= 8 bytes and not a class/struct or container. If it is not possible to implement that specific, then I think we need to do this by hand.

Reply:
Thanks!

I added two quick benchmarks to my benchmark (& testbed) repo here. See the commit here: https://github.com/jmarrec/CppBenchmarks/commit/2444a051ac51844c66bd41e3510f9e946610d322

It includes the results of running the benchmarks and the alrgorithmic complexity of each solution.

`const std::array<double, 2>&` is faster